### PR TITLE
Refactoring questions asked of user

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

In this pull request, I'm extracting a conceptual authorization layer.
This work trackes towards https://github.com/forem/forem/issues/15624.

I want to open this pull request as a conversation around the first step
in the approach.  Namely, moving these role based questions into an
authorization layer.

Ultimately, I want to completely eschew `if user.super_admin?` and favor
something like:

```ruby
if Authorized.for?(user: user, to: :edit, on: @article)
```

Hold that syntax lightly, as it is part of a larger conversation, but
the fundamental issue remains: to understand the permissions of our
application we must avoid inline questions such as
`user.has_role?(:admin)` and instead favor the explicit questions that
ask what is the user trying to do.

Once we've encapsulated that behavior, we can begin to more easily
customize implementations for different Forem installations.

Please note, this pull request is intended to be a conversation.  I need
significant feedback, but am also approaching this having implemented
multiple complex and highly flexible authorization applications.

## Related Tickets & Documents

See #15624.

## QA Instructions, Screenshots, Recordings

None.

### UI accessibility concerns?

None.

## Added/updated tests?

- [x] No, and this is why: this is a draft proposal

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by jeremyf** [comment]
I wrote [Factoring Towards a More Extensible Role System - Forem Team 🌱](https://forem.team/jeremy/factoring-towards-a-more-extensible-role-system-1k90) as a place for internal discussions.

**Comment by msarit** [comment]
Conceptually, this makes sense to me. Being able to query role-related permissions like: `if Authorized.for?(user: user, to: :edit, on: @article)` brings a lot of flexibility that the current role queries do not.

That being said, I am unsure how the `.authorizer` factors into this destination. It makes sense code-wise, though it does feel a little awkward to me (`user.authorizer.has_this_role?`).

**Comment by jeremyf** [comment]
> That being said, I am unsure how the `.authorizer` factors into this destination. It makes sense code-wise, though it does feel a little awkward to me (`user.authorizer.has_this_role?`).

That's great feedback, I'm noodling on this problem as I refactor to "get the authorization code out of the User."  I'm "intuiting" that this step will make future changes easier.  But I think I need to push against it a bit more.  I was hesitant to alter too much of the non-user model logic as part of this conversation.

I want to leave space for others to review and comment.  And reconsider a more concrete step.

**Comment by citizen428** [comment]
This seems a bit fleshed out as a conversation starter and the PR description runs the risk of biasing readers towards an already envisioned solution. I do not question your domain expertise and previous experience in the matter, I just would prefer to be on the right side of this image instead of the left one.

image

This is one of the reasons why we did RFCs in the past. And while I do understand that we wanted to take a step back in this regard this definitely seems like a feature that's big and important enough to warrant one.

File: app/models/user.rb [hunk 1]
```diff
@@ -346,51 +346,95 @@ def cached_antifollowed_tag_names
...
 
-  def trusted
-    ActiveSupport::Deprecation.warn("User#trusted is deprecated, favor User#trusted?")
-    trusted?
-  end
+  # My preference is to go with:
+  #
+  #   `Authorize.for(user: user, to: <action>, on: <subject>)`
+  #
+  # However, this is a refactor, and it's goal is to reduce the direct
```
Comment (citizen428) [inline app/models/user.rb:413]:
```suggestion
  # However, this is a refactor, and its goal is to reduce the direct
```

**Comment by jeremyf** [comment]
Regarding fleshed out versus RFC, as I spent time thinking through what to say to get this moving, expressing this in code made just as much sense.  The work was quick and has already yielded good conversation.

I like the idea of the `Authorizer` factory and can adjust to highlight the approach.  I'm hesitant in this proposed iteration to swap out `user.admin?` for `Authorizer.for(user: user).admin?`.  I really want to limit the impact of this proposed change so that only `User` must account for this new collaboration.  Which, I acknowledge, my `user.authorizer.has_a_role?` does not do.

My plan is to tighten up my intention and update the PR and of the PR.  I will state the benefits and consequences of the change and work to remove any undue "advocacy" for a solution.

**Comment by jeremyf** [comment]
I'm closing this in favor of #15747.  That PR compresses the many changes into a single commit and rebases against upstream changes.

**Pull request closed without merging**